### ISSUE TITLE
spago not working on windows

### ISSUE THREAD

**Issue by Vikasg7** [spacchetti/spago#380]
`spago init` worked fine but `spago build` and `spago install psci-support` are giving the following error:  
`spago.exe: Directory "\\" is not accessible. Permissions {readable = True, writable = False, executable = False, searchable = True}`  
I also tried to run cmd.exe as System Administrator but didn't help. 

purescirpt version: 0.13.2
spago version : 0.9.0.0
windows 8

**Comment by f-f** [comment]
@Vikasg7 thanks for reporting! We'd need some more info in order to debug this.
Could you paste here the full output from `spago -v install`?

Also cc @stkb and @btrepp as you might have more ideas/context than me on this

**Comment by Vikasg7** [comment]
Here it is  
```
Running `spago install`
Transformed config is the same as the read one, not overwriting it
Ensuring that the package set is frozen
Getting transitive deps
Transformed config is the same as the read one, not overwriting it
Running `fetchPackages`
Checking if `purs` is up to date
Running `getGlobalCacheDir`
spago.exe: Directory "\\" is not accessible. Permissions {readable = True, writable = False, executable = False, searchable = True}
```

cc @stkb, @btrepp

**Comment by f-f** [comment]
@Vikasg7 thanks! Some more info:
- what do you get if you run `spago -v install --global-cache=skip`?
- do you have any of the following environment variables set, and if yes what's their value?
  - `HOME`
  - `XDG_CACHE_HOME`
  - `HomePath`

**Comment by Vikasg7** [comment]
> * what do you get if you run `spago -v install --global-cache=skip`?  

Output is same.  

> * do you have any of the following environment variables set, and if yes what's their value?
>   
>   * `HOME`
>   * `XDG_CACHE_HOME`
>   * `HomePath`  

None of the above variables are set. Just a reminder, I am using Windows 8.

**Comment by f-f** [comment]
> None of the above variables are set. Just a reminder, I am using Windows 8.

As far as I know `HomePath` should be set in Windows (but I'm not a Windows user, so I think the fact that you're using Windows 8 vs other versions might be relevant but I don't know how)

Anyways from your output it seems Spago stumbles on getting a place for the global cache, and the logic for that will try to get locations from `XDG_CACHE_HOME`, `HOME` and `HomePath` (in this order) and fallback to the project-local `.spago-global-cache` if all of them are missing (this last option is "safe" because Spago cannot work if it's not possible to create project-local dirs)

Here's the code implementing what I described above:
https://github.com/spacchetti/spago/blob/2efcc8fb84d5dc99cdd6b8c73c68306ce58397b3/src/Spago/GlobalCache.hs#L162-192

**Comment by Vikasg7** [comment]
@f-f   
Which path should I set as value to any of those three variables?  
I just checked again, I don't see any of the above path set in my Environment Variables.

**Comment by f-f** [comment]
@Vikasg7 my point above was that it should not be necessary to set any of them, because if they are not set then Spago will just use a project local folder

However, instead of using `.spago-global-cache` as it should, it tries to use `\\`, which is the weird part

**Comment by Vikasg7** [comment]
@f-f  Yeah, looks like `\\` is not resolving on windows. I set the HOME variable to a local folder and `spago install` worked fine. Thanks for your help.

**Comment by f-f** [comment]
@Vikasg7 you're welcome, happy to hear you solved it 🙂 

Though I still wonder how this could happen in the first place 🤔

**Comment by Vikasg7** [comment]
@f-f  
I dig into this and found out `</>` operator used in https://github.com/spacchetti/spago/blob/2efcc8fb84d5dc99cdd6b8c73c68306ce58397b3/src/Spago/GlobalCache.hs#L166 resolves to `\\` for windows. [Link](http://hackage.haskell.org/package/filepath-1.4.2.1/docs/System-FilePath-Posix.html#v:-60--47--62-) to Haskell documentation. So it makes the path looks like this `.spago-global-cache\\spago`. I tried cding into .spago-global-cache from the project folder but couldn't as this folder doesn't exist.
So the question arises - Does spago really creates `.spago-global-cache` folder before using it?

**Comment by f-f** [comment]
@Vikasg7 thanks for looking into this!

Indeed the folder is not there because the error you got at first was about Spago not being able to create it. The error you get comes from here:
https://github.com/spacchetti/spago/blob/64998a7944a87cba0bfec472aaf0fec411c0e6d1/src/Spago/FetchPackage.hs#L49-L51

The part that creates the directory is the `assertDirectory`, but in your case it fails, in this part (because the `\\` already exists):
https://github.com/spacchetti/spago/blob/64998a7944a87cba0bfec472aaf0fec411c0e6d1/src/Spago/Prelude.hs#L235-L239

So in the end I think the problem happens in the line of code you linked in the last message. Could you try replacing this line
https://github.com/spacchetti/spago/blob/64998a7944a87cba0bfec472aaf0fec411c0e6d1/src/Spago/GlobalCache.hs#L152

with the following:

```hs
let result = cacheDir </> "spago"
echoStr result
pure result
```

So we should be able to see what happens?

**Comment by Vikasg7** [comment]
@f-f  
So, I install haskell and stack on my pc, git cloned the repo, did the changes and tried `stack build --fast` but got some error. Is this is how I should go about it? or you have some better way to try out the changes? Here is the error though.  
```
 Building library for network-2.6.3.6..
    [1 of 9] Compiling Network.Socket.ByteString.Internal ( Network\Socket\ByteString\Internal.hs, .stack-work\dist\7d103d30\build\Network\Socket\ByteString\Internal.o )
    [2 of 9] Compiling Network.Socket.Types ( .stack-work\dist\7d103d30\build\Network\Socket\Types.hs, .stack-work\dist\7d103d30\build\Network\Socket\Types.o )

    C:\Users\1622\AppData\Local\Temp\stack3700\network-2.6.3.6\Network\Socket\Types.hsc:764:16: error:
        parse error on input `CALLCONV'
        |
    764 | foreign import CALLCONV unsafe "ntohs" ntohs :: Word16 -> Word16
```

**Comment by stkb** [comment]
This happens because the `HomePath` variable checked for does not contain a drive letter; ie it is just `\Users\{username}`. So the you're asking Windows to create a folder `\Users\{username}\.cache`. I'll submit a PR to fix this shortly (I've had it on the shelf for a while)

**Comment by f-f** [comment]
@Vikasg7 this is a known problem of the set of packages we're using, see [here](https://github.com/commercialhaskell/stackage/issues/3385) for more details on it

We had to work around it on our Windows CI too, in this way:
https://github.com/spacchetti/spago/blob/13cf90cddb1e3e2b7338169d98799b6bc6a54ffe/appveyor.yml#L11-L14

**Comment by Vikasg7** [comment]
@sktb, @f-f  
I did `spago run -w` and got this
```
Installation complete.
CreateFile "E:\\Desktop\\Purescript\\prac\\Desktop\\Purescript\\prac\\.spago\\bifunctors\\v4.0.0\\src\\": does not exist (The system cannot find the path specified.)
```  
As you can see its prepending an absulute path to my Desktop to a relative path to the target folder. Is this issue also getting fixed with above patch?  
If yes, how can I install the latest git binaries globally, I tried it through github via npm but it says package.json doesn't exist.  
 `spago run` works fine though.

**Comment by f-f** [comment]
@Vikasg7 the npm install works only on tags, so the fix will ship with the next release (which we could cut in the next days)

**Comment by Vikasg7** [comment]
@f-f   
Can you confirm `spago run -w` work with the latest commit? I think this issue is also related to paths.

**Comment by f-f** [comment]
@Vikasg7 yes, you can take a look at the test results in the Windows CI run for the latest commit on master [here](https://ci.appveyor.com/project/f-f/spago/builds/26936884): there are two tests for `spago run` (line 267) and they both pass

**Comment by stkb** [comment]
@Vikasg7 @f-f  I can repro the same issue.

While `spago run` works fine, adding the `-w` for the watch function has some (other) issue with paths. (I never use the watch function, so I never came across this before)

This shouldn't be too difficult to fix though I guess.

**Comment by stkb** [comment]
@Vikasg7 Thanks for finding that issue (it was something completely separate). Watching on Windows should now be fixed in (#387).
